### PR TITLE
WEB-622. show invalid simulations

### PR DESCRIPTION
**Pull request by TamoshaytisV**
https://liv-it.atlassian.net/browse/WEB-622

The reason of the bug was invalid LTI URL configured in studio. To prevent this case validation was added to existing labster logic which defines which simulations have errors in url and displays its list

@polesye @leha2000 please review

### CODE REVIEW

File: lms/djangoapps/labster_course_license/utils.py [hunk 1]
```diff
@@ -105,8 +106,13 @@ def course_tree_info(store, simulations, licensed_simulations):
     Retuns information about the course's xblocks.
     """
     course_info = {}
+    invalid_simulations = []
+    validation_pattern = r'^https:\/\/[\w\d]*\.labster\.com\/[a-z]*\/[\w\d]*\/$'
```
Comment (polesye) [inline lms/djangoapps/labster_course_license/utils.py]: does it make sense to validate entire url or, probably, simulation id will be enough?

Comment (polesye) [inline lms/djangoapps/labster_course_license/utils.py]: it looks like we're using `[a-zA-Z0-9]+` and you cannot use `*`. See https://github.com/Livit/Livit.Learn.LTI/blob/master/django_project/core/urls.py#L12

File: lms/djangoapps/labster_course_license/utils.py [hunk 2]
```diff
@@ -105,8 +106,13 @@ def course_tree_info(store, simulations, licensed_simulations):
...
     for simulation in simulations:
-        simulation_id = get_simulation_id(simulation.launch_url)
+        lti_url = simulation.launch_url
+        simulation_id = get_simulation_id(lti_url)
+        if not simulation_id or not re.search(validation_pattern, lti_url):
+            invalid_simulations.append(simulation.display_name)
```
Comment (polesye) [inline lms/djangoapps/labster_course_license/utils.py]: do we need to skip all the further steps if simulation id is wrong?

Comment (polesye) [inline lms/djangoapps/labster_course_license/utils.py]: probably `continue`.

Comment (polesye) [inline lms/djangoapps/labster_course_license/utils.py]: `invalid_simulations` bad naming. Simulations are okay, wrong simulation ids :)

File: lms/djangoapps/labster_course_license/views.py [hunk 3]
```diff
@@ -16,7 +16,7 @@
 from django.shortcuts import redirect
 from django.views.decorators.cache import cache_control
 from django.views.decorators.csrf import ensure_csrf_cookie
-from django.http import HttpResponseBadRequest
+from django.http import HttpResponseBadRequest, Http404
```
Comment (polesye) [inline lms/djangoapps/labster_course_license/views.py:19]: why do we need `Http404`?

File: lms/djangoapps/labster_course_license/views.py [hunk 4]
```diff
@@ -215,7 +215,15 @@ def set_license(request, course, ccx):
     store = modulestore()
     with store.bulk_operations(course_key):
         simulations = store.get_items(course_key, qualifiers={'category': 'lti'})
-        course_info, chapters = course_tree_info(store, simulations, licensed_simulations)
+        course_info, chapters, invalid_simulations = course_tree_info(
```
Comment (polesye) [inline lms/djangoapps/labster_course_license/views.py]: `invalid_simulations` bad naming.

File: lms/djangoapps/labster_course_license/views.py [hunk 5]
```diff
@@ -215,7 +215,15 @@ def set_license(request, course, ccx):
...
+            store, simulations, licensed_simulations
+        )
+        if invalid_simulations:
+            messages.error(
+                request,
+                _('Validation error. Please check urls for following simulations {}.'.format(str(invalid_simulations)))
```
Comment (polesye) [inline lms/djangoapps/labster_course_license/views.py]: Please rewrite the message in th following way:

``` python
'Please verify LTI URLs  are coorrect for the following simulations: \n {}'.format(
     '\n'.join(
            'name: "{}", simulation Id: "{}"'.format(x.name, x.id) fox x in y
      ...
)
```

Result:

```
Please verify LTI URLs  are coorrect for the following simulations:
name: 'Adaptive Learning Simulation', simulation id: 'ASGFDA123123  '
name: 'MDL, simulation id: '  AS G F D A 1asd 23123  ',
```

**Comment by polesye** [comment]
could you add tests?

File: lms/djangoapps/labster_course_license/utils.py (hunk 1 above)
Comment (TamoshaytisV) [inline lms/djangoapps/labster_course_license/utils.py]: > does it make sense to validate entire url or, probably, simulation id will be enough?

It makes sense 'cause we can have tabs/spaces/other chars before/after url

File: lms/djangoapps/labster_course_license/utils.py (hunk 2 above)
Comment (TamoshaytisV) [inline lms/djangoapps/labster_course_license/utils.py]: > invalid_simulations bad naming. Simulations are okay, wrong simulation ids :)

Not ids :) I store simulation names that have invalid lti url, renamed to `invalid_simulation_names`
@polesye maybe you have better suggestion

File: lms/djangoapps/labster_course_license/views.py (hunk 3 above)
Comment (TamoshaytisV) [inline lms/djangoapps/labster_course_license/views.py:19]: > why do we need Http404?

it's used on line 161 but was not imported so it was potential exception

File: lms/djangoapps/labster_course_license/utils.py (hunk 1 above)
Comment (polesye) [inline lms/djangoapps/labster_course_license/utils.py]: > It makes sense 'cause we can have tabs/spaces/other chars before/after url

but this particular functionality works only with `simulation_id`. And the problems were with them. 
I believe that LTI module cut all whitespaces at the beginnig and at the end before save, but you have to double check that.
Also, I don't like the idea of hardcoding protocol and labster.com here. Probably we can check for white spaces, tabulations, etc..

Comment (polesye) [inline lms/djangoapps/labster_course_license/utils.py]: in this case it doesn't work with local enviroment.
I'd suggest to use 2 validations:
1. Validate general url format.
2. Validate simulation id.

**Comment by polesye** [comment]
@TamoshaytisV any updates on that?

File: lms/djangoapps/labster_course_license/utils.py (hunk 1 above)
Comment (leha2000) [inline lms/djangoapps/labster_course_license/utils.py]: You could split by `'/simulation/'` string and get the server string and `simulation_id + '/'`

**Comment by TamoshaytisV** [comment]
> @TamoshaytisV any updates on that?

@polesye working on tests

File: lms/djangoapps/labster_course_license/tests/test_views.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+            data={'license': self.license, 'update': True}
+        )
+        request.user = self.instructor
+        self.create_valid_simulations()
+        res = set_license(request, self.course, self.ccx)
+        messages = res.context.get('messages')[0]
+        ccx_locator = CCXLocator.from_course_locator(self.course.id, self.ccx.id)
+        url = reverse('labster_license_handler', kwargs={'course_id': ccx_locator})
+        self.assertEqual(res.status_code, 302)
+        self.assertRedirects(res, url)
```
Comment (polesye) [inline lms/djangoapps/labster_course_license/tests/test_views.py]: where can I find test for wrong ids?

File: lms/djangoapps/labster_course_license/views.py [hunk 7]
```diff
@@ -215,7 +215,23 @@ def set_license(request, course, ccx):
...
     with store.bulk_operations(course_key):
         simulations = store.get_items(course_key, qualifiers={'category': 'lti'})
-        course_info, chapters = course_tree_info(store, simulations, licensed_simulations)
+        course_info, chapters, invalid_simulation_ids = course_tree_info(
+            store, simulations, licensed_simulations
+        )
+        if invalid_simulation_ids:
+            messages.error(
+                request,
+                _((
```
Comment (polesye) [inline lms/djangoapps/labster_course_license/views.py]: please move it to `msg` variable.

**Comment by polesye** [comment]
@TamoshaytisV please add a screenshot with results

**Comment by TamoshaytisV** [comment]
screen shot 2016-06-20 at 10 12 25 pm

@polesye @leha2000 please continue review

File: openedx/core/djangoapps/labster/tests/base.py [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+"""
+from xmodule.modulestore.tests.factories import CourseFactory
+from xmodule.modulestore.tests.django_utils import ModuleStoreTestCase, TEST_DATA_SPLIT_MODULESTORE
+from ccx.tests.factories import CcxFactory
+from ccx_keys.locator import CCXLocator
+from student.roles import CourseCcxCoachRole
+from student.tests.factories import UserFactory
+
+
+class LTITestBase(ModuleStoreTestCase):
```
Comment (polesye) [inline openedx/core/djangoapps/labster/tests/base.py]: is it really about LTI? I think it's about CCX coarses

**Comment by avkoval** [comment]
:+1:  (but we have ideas how to improve it on next round)

**Comment by polesye** [comment]
👍

**Pull request merged**